### PR TITLE
fix: Support internal endpoints

### PR DESCRIPTION
**Pull request by tolusha**
Signed-off-by: Anatolii Bazko <abazko@redhat.com>

### What does this PR do?
Support internal endpoints

### Screenshot/screencast of this PR
Devfile: https://gist.githubusercontent.com/l0rd/baa74d544f38c7fd358d1711dd46a071/raw/exposures-devfile.yaml
Exposed endpoints: 
screenshot-console-openshift-console apps ci-ln-g2x9hyk-72292 origin-ci-int-gce dev rhcloud com-2022 05 12-16_02_32

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/21387

### How to test this PR?

https://github.com/eclipse-che/che-operator/pull/1384#issuecomment-1137823573

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Custom resource definition file is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#updating-custom-resource-definition-file)
- [ ] [OLM bundles are up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-olm-bundle)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Review by amisevsk — COMMENTED** [review]
I'm not sure the current changes are sufficient -- when I tested this PR via the golang sample, setting the `8080` endpoint to `internal`, I was still able to access it via the public endpoint (meaning it was treated as `public`). 

I think the different exposures should be treated as follows:
* `none` exposure: no service or route is created, endpoint exists for information/configuration purposes
* `internal` exposure: only a service is created, endpoint is not accessible from outside the cluster (i.e. you can't open the endpoint in Theia). Endpoint attribute `controller.devfile.io/endpoint-url` is set to the service hostname
* `public` exposure: a route/ingress is created or the gateway is set up to serve this route, endpoint is accessible from the internet, and endpoint attribute is set to public URL.

**Review by amisevsk — COMMENTED** [review]
I've tested these changes with the following devworkspace applied directly to the cluster:

<details>
  <summary>devworkspace.yaml</summary>
  
```yaml
apiVersion: workspace.devfile.io/v1alpha2
kind: DevWorkspace
metadata:
  name: test-che
spec:
  routingClass: che
  started: true
  template:
    attributes:
      controller.devfile.io/storage-type: ephemeral
    components:
      - name: redis-none
        container:
          image: redis:latest
          endpoints:
            - name: redis-none
              attributes:
                urlRewriteSupported: true
                cookiesAuthEnabled: true
              targetPort: 6380
              exposure: none
          args: ['redis-server', '--port', '6380']

      - name: redis-internal
        container:
          image: redis:latest
          endpoints:
            - name: redis-int
              attributes:
                urlRewriteSupported: true
                cookiesAuthEnabled: true
              targetPort: 6381
              exposure: internal
          args: ['redis-server', '--port', '6381']

      - name: redis-public
        container:
          image: redis:latest
          endpoints:
            - name: redis-pub
              attributes:
                urlRewriteSupported: true
                cookiesAuthEnabled: true
              targetPort: 6382
              exposure: public
          args: ['redis-server', '--port', '6382']

      - name: redis-none-no-rewrite
        container:
          image: redis:latest
          endpoints:
            - name: redis-none-nr
              targetPort: 6383
              exposure: none
          args: ['redis-server', '--port', '6383']

      - name: redis-internal-no-rewrite
        container:
          image: redis:latest
          endpoints:
            - name: redis-int-nr
              targetPort: 6384
              exposure: internal
          args: ['redis-server', '--port', '6384']

      - name: redis-public-no-rewrite
        container:
          image: redis:latest
          endpoints:
            - name: redis-pub-nr
              targetPort: 6385
              exposure: public
          args: ['redis-server', '--port', '6385']
```
</details>

The purpose of this file is to test the 6 combinations of settings for endpoints:

* An endpoint can have exposure `none`, `internal`, or `public`
* Endpoints can have the `urlRewriteSupported: true` attribute to expose them through the che-gateway component (this is used for e.g. the Theia editor)

I found that the Che Operator correctly assigns endpoint URLs as expected:
```bash
> oc get cm workspace096af80eedf14c47-metadata -o yaml | yq '.data."flattened.devworkspace.yaml"' -r | yq -y '.components[].container?.endpoints[] | {"name": .name, "url": .attributes."controller.devfile.io/endpoint-url"}'
name: redis-none
url: null
---
name: redis-int
url: http://workspace096af80eedf14c47-service.user1-che.svc:6381
---
name: redis-pub
url: https://eclipse-che.apps.<snip>.com/workspace096af80eedf14c47/redis-public/6382/
---
name: redis-none-nr
url: null
---
name: redis-int-nr
url: http://workspace096af80eedf14c47-service.user1-che.svc:6384
---
name: redis-pub-nr
url: http://workspace096af80eedf14c47-1.apps.<snip>.com/
```

However, looking at the routing objects actually created on the cluster, there are some issues:
* There is a route created for the `redis-int-nr`, exposing it to the internet despite it being internal (the Che Operator should not create a route for this service)
* While there is no officially-created route for the `redis-int` endpoint, the che-gateway component is configured to forward requests along that port

In short, the Che Operator is treating `internal` exposure endpoints as `internal` in terms of what it reports to the user, but is setting up the networking as though those endpoints have a `public` exposure.

As an aside, it looks like the redis server just returns a 503 on access -- does anyone know a simple server that cna be configured to listen on any port?

**Review by amisevsk — APPROVED** [review]
LGTM

If cluster-bot comes back to life I'll test it again, but I don't see any issues.

**Comment by tolusha** [comment]
Test added
